File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day08Matchsticks.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day08Matchsticks(filename: String) extends DailyProblem[Int, Int] {

  private val input = FileUtilities.readFile(filename)

  @tailrec
  final def encode(input: String, working: String): String = {
    if (input.isEmpty) {
      "\"" + working + "\""
    } else {
      encode(input.tail, working + (input.head match {
        case '\\' => "\\\\"
        case '"' => "\\\""
        case c => c.toString
      }))
    }
  }

  @tailrec
  final def decode(input: String, working: String): String = {
    if (input.isEmpty) {
      working
    } else if (input.startsWith("\"")) {
      decode(input.tail, working)
    } else if (input.startsWith("\\\\")) {
      decode(input.drop(2), working + "\\")
    } else if (input.startsWith("\\\"")) {
      decode(input.drop(2), working + "\"")
    } else if (input.startsWith("\\x")) {
      val hex = Integer.parseInt(input.slice(2, 4), 16).intValue.toChar
      decode(input.drop(4), working + hex)
    } else {
      decode(input.tail, working + input.head)
    }
  }

  private val inputLength = input.map(line => line.length).sum
  private val encodedLength = input.map(line => encode(line, "").length).sum
  private val decodedLength = input.map(line => decode(line, "").length).sum

  override val part1Answer: Int = inputLength - decodedLength
  override val part2Answer: Int = encodedLength - inputLength
}

File: java-advent-of-code/src/main/java/advent/twenty_fifteen/Day08Matchsticks.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

import java.util.List;
import java.util.stream.Collectors;

public class Day08Matchsticks implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    public static String decode(String line) {
        line = line.substring(1, line.length() - 1);
        String result = "";
        for (int i = 0; i != line.length(); i++) {
            if (line.charAt(i) == '\\') {
                i++;
                switch (line.charAt(i)) {
                    case '\\':
                        result += "\\";
                        break;
                    case '\"':
                        result += "\"";
                        break;
                    case 'x':
                        i++;
                        char hex0 = line.charAt(i);
                        i++;
                        char hex1 = line.charAt(i);
                        result += (char) Integer.parseInt("" + hex0 + hex1, 16);
                        break;
                    default:
                        throw new IllegalArgumentException("Unknown escape character (" + line.charAt(i) + ")");
                }
            } else {
                result += line.charAt(i);
            }
        }
        return result;
    }

    public static String encode(String line) {
        String result = "";

        for (char c : line.toCharArray()) {
            if (c == '\\') {
                result += "\\\\";
            } else if (c == '\"') {
                result += "\\\"";
            } else {
                result += c;
            }
        }
        return "\"" + result + "\"";
    }

    public Day08Matchsticks(String filename) {
        List<String> raw = FileUtilities.readLines(filename, Parsers::TO_STRING);
        List<String> decoded = raw.stream().map(s -> decode(s)).collect(Collectors.toList());
        List<String> encoded = raw.stream().map(s -> encode(s)).collect(Collectors.toList());

        int rawSpace = raw.stream().mapToInt(s -> s.length()).sum();
        int decodedSpace = decoded.stream().mapToInt(s -> s.length()).sum();
        int encodedSpace = encoded.stream().mapToInt(s -> s.length()).sum();

        this.part1Answer = rawSpace - decodedSpace;
        this.part2Answer = encodedSpace - rawSpace;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}
